File: Java/src/main/java/com/DSA/Java/Graphs/TopologicalSort/TpsUsingList.java
package com.DSA.Java.Graphs.TopologicalSort;
import com.DSA.Java.Graphs.Traversal.BFSAndDFS.AdjacencyList.*;
import java.util.Stack;

public class TpsUsingList extends Graph {
    private void tpsHelper(GraphNode node, Stack<GraphNode> stack) {
        for (GraphNode tempNode : node.neighbours) {
            if (!tempNode.isVisisted) {
                tpsHelper(tempNode, stack);
            }
        }
        node.isVisisted = true;
        stack.push(node);
    }

    public void tps() {
        Stack<GraphNode> stack = new Stack<>();
        for (GraphNode node : getNodeList()) {
            if (!node.isVisisted) {
                tpsHelper(node, stack);
            }
        }

        while (!stack.isEmpty()) {
            System.out.print(stack.pop().name + " ");
        }
        System.out.println();
    }
}

File: C++/Graph/src/TopologicalSort/TpsUsingList.cpp
#include "../../include/TopologicalSort/TpsUsingList.h"

TpsUsingList::TpsUsingList() {
  AdjacencyList();
}

void TpsUsingList::tps_helper(GraphNode *node) {
  for (const auto &node : node->neighbours) {
    if (!node->is_visited) {
      tps_helper(node);
    }
  }

  node->is_visited = true;
  stack.push(node);
}

void TpsUsingList::tps() {
  std::vector<GraphNode *> node_list = get_node_list();
  for (const auto &node : node_list) {
    if (!node->is_visited) {
      tps_helper(node);
    }
  }

  std::cout << "\n";
  while (!stack.empty()) {
    GraphNode *current_node = stack.top();
    stack.pop();
    std::cout << current_node->name << " ";
  }
  std::cout << "\n";
}
